src/content/advanced/connectivity/ingress/service-type-loadbalancer/index.md

```diff
@@ -229,41 +262,54 @@ To avoid downtime, we can create an additional Kubernetes `Service` of type `Loa
-    # value. Defaults to 5, must be between 2 and 60.
-    service.beta.kubernetes.io/aws-load-balancer-extra-security-groups: "sg-53fae93f,sg-42efd82e"
-    # A list of additional security groups to be added to the ELB.
+    # Disable AWS Network Load Balancer client IP preservation.
+    service.beta.kubernetes.io/aws-load-balancer-target-group-attributes: preserve_client_ip.enabled=false
 ```
 
+See [Target groups for your Network Load Balancers: Client IP preservation](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/load-balancer-target-groups.html#client-ip-preservation) for more information about this whole feature.
+
+##### Health Checks failing when using PROXY protocol and `externalTrafficPolicy: Local`
+
+The before mentioned limitation directly leads us the next pitfall: One could think "well, if the integrated client IP preservation is not working, I can still use PROXY protocol". In theory and at least for the Kubernetes integrated Cloud Controller this should work. In theory.
+
+In reality we need to step back and take a look at how health checks are being implemented with `externalTrafficPolicy: Local`: By default and with `externalTrafficPolicy: Cluster` the AWS Network Load Balancer sends its health check requests to the same port it's sending traffic to: The traffic ports defined in the Kubernetes service. From there they are getting answered by the pods backing your service.
+
+Now, when enabling PROXY protocol, AWS assumes your service is able to understand PROXY protocol for both, traffic and health checks. And this is what happens for services using `externalTrafficPolicy: Cluster` with PROXY protocol enabled: AWS is using it for both the traffic and the health check because in the end and if not configured otherwise by using annotations, they end up on the same port.
+
+But things change when using `externalTrafficPolicy: Local`: Local means the traffic hitting a node on the allocated traffic node port stays on the same node. Therefore only nodes running at least one pod of your service are eligible targets for the AWS Network Load Balancer's target group as other nodes fail to respond to its health checks.
+
+Since the health check might get false negative when two pods are running on the same node and one of them is not healthy (anymore), Kubernetes allocates a separate health check node port and configures it in the AWS Network Load Balancer. This health check node port and requests hitting it is handled by `kube-proxy` or its replacement (Cilium in our case). Unfortunately both of them are not able to handle PROXY protocol and therefore all health checks will fail starting the moment you enable PROXY protocol in your AWS Network Load Balancer in conjunction with `externalTrafficPolicy: Local`.
+
+At last this means there is currently no way of preserving the original client IP using internal AWS Network Load Balancers being accessed from inside the same cluster.
```

Comment:
> At last this means there is currently no way of preserving the original client IP using internal AWS Network Load Balancers being accessed from inside the same cluster.

With `externalTrafficPolicy: Cluster`, proxy protocol works fine, doesn't it?

Reply:
As far as I remember, yes, because the health check is being sent via the traffic ports and Ingress NGINX Controller (or whatever is lying behind) can handle this.

Reply:
So then the paragraph is not accurate :)

Reply:
> It is accurate, because externalTrafficPolicy: Cluster does not preserve the IP address as it always uses SNAT.

Does this mean that it's not possible in any way to preserve the client IP using NLBs and Cilium? I was under the impression that having NLB with Proxy protocol and `externalTrafficPolicy: Cluster` it does preserve the client IP.

Reply:
I think we have a different understanding of "preserving the client IP". What I mean is not changing the client IP at all and just passing packets without changing the source address (layer 3), so the node sees the client IP and thinks it's routing traffic back directly to the client (it isn't, due to AWS network and NLB intercepting it).

What you mean is keeping track of the client IP by injecting it into the TCP connection (layer 4) using PROXY protocol.

The latter is not supported for NLBs managed by AWS Load Balancer Controller and `externalTrafficPolicy: Local` as the daemon exposing the `healthCheckNodePort` for `externalTrafficPolicy: Local` would need to understand PROXY protocol, but both `kube-proxy` and Cilium are not doing so.

So yes, you could also just fallback to using NLBs with NLB Client IP Preservation (my understanding, layer 3) turned off and use PROXY protocol (your understanding, layer 4) with `externalTrafficPolicy: Cluster` instead. But this would also add an extra hop and make each and every node eligible for traffic, even when not running an endpoint of the service.

Reply:
Love it. Thank you!